plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeLocalConcurrentWritesTest.java

```diff
@@ -0,0 +1,442 @@
+            // T1: (1)
+            // T2: (2)
+            // T3: (3)
+            List<Future<Boolean>> futures = IntStream.range(0, threads)
+                    .mapToObj(threadNumber -> executor.submit(() -> {
+                        barrier.await(10, SECONDS);
+                        try {
+                            getQueryRunner().execute("INSERT INTO " + tableName + " SELECT COUNT(*) FROM " + tableName);
+                            return true;
+                        }
+                        catch (Exception e) {
+                            RuntimeException trinoException = getTrinoExceptionCause(e);
+                            try {
+                                assertThat(trinoException).hasMessage("Failed to write Delta Lake transaction log entry");
+                            }
+                            catch (Throwable verifyFailure) {
+                                if (verifyFailure != e) {
+                                    verifyFailure.addSuppressed(e);
+                                }
+                                throw verifyFailure;
+                            }
+                            return false;
+                        }
+                    }))
+                    .collect(toImmutableList());
+
+            long successfulInsertsCount = futures.stream()
+                    .map(future -> tryGetFutureValue(future, 20, SECONDS).orElseThrow(() -> new RuntimeException("Wait timed out")))
+                    .filter(success -> success)
+                    .count();
+
+            assertThat(successfulInsertsCount).isGreaterThanOrEqualTo(1);
+            assertQuery(
+                    "SELECT * FROM " + tableName,
+                    "VALUES (0)" +
+                            LongStream.rangeClosed(1, successfulInsertsCount)
+                                    .boxed()
+                                    .map("(%d)"::formatted)
+                                    .collect(joining(", ", ", ", "")));
+            assertQuery(
+                    "SELECT version, operation, isolation_level, read_version FROM \"" + tableName + "$history\"",
+                    "VALUES (0, 'CREATE TABLE AS SELECT', 'WriteSerializable', 0)" +
+                            LongStream.rangeClosed(1, successfulInsertsCount)
+                                    .boxed()
+                                    .map(version -> "(%s, 'WRITE', 'WriteSerializable', %s)".formatted(version, version - 1))
+                                    .collect(joining(", ", ", ", "")));
+        }
+        finally {
+            assertUpdate("DROP TABLE " + tableName);
+            executor.shutdownNow();
+            assertTrue(executor.awaitTermination(10, SECONDS));
+        }
+    }
+
+    @Test
+    public void testConcurrentMixedInsertsSelectingFromTheSameTableAndBlindInserts()
+            throws Exception
+    {
+        int threads = 3;
```

Comment:
@alexjo2144 FYI added here (based on our offline discussion) a testcase to showcase the scenario where the `INSERT` operation doing `SELECT` from the same table does not collide with blind inserts in `WriteSerializable` isolation level.